File: src/main/kotlin/com/linqingying/cangjie/psi/CjEscapeStringTemplateEntry.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.psi.stubs.CangJiePlaceHolderWithTextStub;
import com.intellij.lang.ASTNode;
import com.intellij.openapi.util.text.StringUtil;
import org.jetbrains.annotations.NotNull;

import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes;
import org.jetbrains.annotations.Nullable;

public class CjEscapeStringTemplateEntry extends CjStringTemplateEntry {
    public CjEscapeStringTemplateEntry(@NotNull ASTNode node) {
        super(node);
    }

    public CjEscapeStringTemplateEntry(@NotNull CangJiePlaceHolderWithTextStub<CjEscapeStringTemplateEntry> stub) {
        super(stub, CjStubElementTypes.ESCAPE_STRING_TEMPLATE_ENTRY);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitEscapeStringTemplateEntry(this, data);
    }

    public String getUnescapedValue() {
        return StringUtil.unescapeStringCharacters(getText());
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjEscapeStringTemplateEntry.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.psi.stubs.CangJiePlaceHolderWithTextStub
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode
import com.intellij.openapi.util.text.StringUtil

class CjEscapeStringTemplateEntry : CjStringTemplateEntry {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJiePlaceHolderWithTextStub<CjEscapeStringTemplateEntry >) : super(
        stub,
        CjStubElementTypes.ESCAPE_STRING_TEMPLATE_ENTRY
    )

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitEscapeStringTemplateEntry(this, data)
    }

    val unescapedValue: String
        get() = StringUtil.unescapeStringCharacters(text)
}
